ngraph/core/reference/src/runtime/reference/tile.cpp

```diff
@@ -14,34 +14,99 @@
+    /// and for 3 (second axis) it would be (4)
+    /// If dims has shape(2, 3, 4) then the output vector would be (3 * 4, 4, 1)
+    /// The outermost axis is not used. For innermost axis it is always 1.
+    /// \param[in] dims Shape of the output
+    ///
+    /// \return Vector containing calculated values for each axis.
+    std::vector<int64_t> create_pitches(const Shape& dims)
+    {
+        std::vector<int64_t> pitch;
+        pitch.resize(dims.size() - 1);
+        std::partial_sum(
+            dims.rbegin(), dims.rend() - 1, pitch.rbegin(), std::multiplies<int64_t>());
+        pitch.push_back(1);
+        return pitch;
+    }
+
+    bool calculate_next_axis(std::vector<int64_t>& indices, size_t& axis, const Shape& shape)
+    {
+        if (axis-- == 0)
+        {
+            return false;
+        }
+
+        if (++indices[axis] != shape[axis])
+        {
+            axis = indices.size();
+            return false;
+        }
+
+        indices[axis] = 0;
+        return true;
+    }
+}
+
+void runtime::reference::tile(const char* arg,
+                              char* out,
+                              const Shape& in_shape,
+                              const Shape& out_shape,
+                              const size_t elem_size,
+                              const std::vector<int64_t>& repeats)
 {
     Shape in_shape_expanded(in_shape);
     in_shape_expanded.insert(in_shape_expanded.begin(), out_shape.size() - in_shape.size(), 1);
-    CoordinateTransform input_transform(in_shape_expanded);
-    CoordinateTransform output_transform(out_shape);
+    size_t block_size = 0;
+    int64_t num_repeats = 0;
+    const int input_rank = in_shape_expanded.size();
+    const int64_t last_dim = in_shape_expanded[input_rank - 1];
+    const std::vector<int64_t> pitches = create_pitches(out_shape);
+    const char* copy = nullptr;
 
-    for (const Coordinate& output_coord : output_transform)
+    std::vector<int64_t> indices(in_shape_expanded.size() - 1, 0);
+    size_t axis = indices.size();
+
+    while (axis <= indices.size())
     {
-        std::vector<size_t> coord;
-        for (auto i = 0; i < output_coord.size(); i++)
+        block_size = last_dim * elem_size;
+        memcpy(out, arg, block_size);
+        out += block_size;
+        arg += block_size;
+
+        copy = out - block_size;
+        num_repeats = repeats[input_rank - 1] - 1;
+        for (int64_t i = 0; i < num_repeats; ++i)
         {
-            auto val = output_coord[i] % in_shape_expanded[i];
-            coord.push_back(val);
+            memcpy(out, copy, block_size);
+            out += block_size;
         }
-        Coordinate input_coord(coord);
 
-        memcpy(out + output_transform.index(output_coord) * elem_size,
-               arg + input_transform.index(input_coord) * elem_size,
-               elem_size);
+        while (calculate_next_axis(indices, axis, in_shape_expanded))
+        {
```

Comment:
I think we can change it to:
```suggestion
        for (; axis != 0; --axis)  // or while (axis-- != 0)
        {
            if (++indices[axis] != shape[axis])
            {
                axis = indices.size();
                break;
            }
            indices[axis] = 0;
```

Reply:
Also if we change first cycle to `do {} while()` we can use the next conditional `axis != 0`.